### PR TITLE
Add strafe quality meter

### LINKED ISSUE #268: Add angle accuracy meter
Angle accuracy meter displays how close to perfect angles your angles were over time.

### PR DESCRIPTION
**Pull request by suburbski**
Closes #268

### CODE REVIEW

**Comment by isRyven** [comment]
Looks good code wise. I liked the way you have relayed reset command through trap, interesting approach to keep the module isolated, although this probably will be executed only on the next frame. I didn't test the feature though.

File: src/cgame/cg_main.cpp [hunk 1]
```diff
@@ -1009,6 +1019,14 @@ cvarTable_t cvarTable[] =
 	{ &etj_jumpSpeedsShowDiff, "etj_jumpSpeedsShowDiff", "0", CVAR_ARCHIVE },
 	{ &etj_jumpSpeedsFasterColor, "etj_jumpSpeedsFasterColor", "0.0 1.0 0.0 1.0", CVAR_ARCHIVE },
 	{ &etj_jumpSpeedsSlowerColor, "etj_jumpSpeedsSlowerColor", "1.0 0.0 0.0 1.0", CVAR_ARCHIVE },
+	// Strafe quality
+	{ &etj_drawStrafeQuality, "etj_drawStrafeQuality", "1", CVAR_ARCHIVE },
+	{ &etj_strafeQualityX, "etj_strafeQualityX", "100", CVAR_ARCHIVE },
+	{ &etj_strafeQualityY, "etj_strafeQualityY", "100", CVAR_ARCHIVE },
```
Comment (Aciz) [inline src/cgame/cg_main.cpp:1025]: I'd rather have these as relative than absolute values, I know it's nowhere consistent in the codebase but relative values make sure they are by default on same place regardless of screen aspect ratio.

File: src/cgame/etj_cgaz.cpp [hunk 2]
```diff
@@ -537,4 +537,130 @@ namespace ETJump
...
 	}
+
+	bool strafingForwards(const playerState_t& ps)
+	{
+		// "forwards" means player viewangles naturally match keys pressed, i.e. looking
+		// ahead with +forward and looking backwards with +back, except for nobeat:
+		// looking to the left with +forward only and looking to the right with +back only
+
+		// get player speed
+		const float speed = std::sqrt(ps.velocity[0] * ps.velocity[0] + ps.velocity[1] * ps.velocity[1]);
```
Comment (Aciz) [inline src/cgame/etj_cgaz.cpp:548]: You can use `VectorLengthSquared2(ps.velocity)` here to make it a bit simpler.

File: src/cgame/etj_cgaz.cpp [hunk 3]
```diff
@@ -537,4 +537,130 @@ namespace ETJump
...
+			}
+		}
+
+		return false;
+	}
+
+	float getOptAngle(const playerState_t& ps)
+	{
+		// get player speed
+		const float speed = std::sqrt(ps.velocity[0] * ps.velocity[0] + ps.velocity[1] * ps.velocity[1]);
```
Comment (Aciz) [inline src/cgame/etj_cgaz.cpp:612]: Same here, `VectorLengthSquared2(ps.velocity)`.

**Review by Aciz — COMMENTED** [review]
Looking good, I will test this next week when I'm home.

File: src/cgame/cg_main.cpp [hunk 4]
```diff
@@ -1009,6 +1019,14 @@ cvarTable_t cvarTable[] =
 	{ &etj_jumpSpeedsShowDiff, "etj_jumpSpeedsShowDiff", "0", CVAR_ARCHIVE },
 	{ &etj_jumpSpeedsFasterColor, "etj_jumpSpeedsFasterColor", "0.0 1.0 0.0 1.0", CVAR_ARCHIVE },
 	{ &etj_jumpSpeedsSlowerColor, "etj_jumpSpeedsSlowerColor", "1.0 0.0 0.0 1.0", CVAR_ARCHIVE },
+	// Strafe quality
+	{ &etj_drawStrafeQuality, "etj_drawStrafeQuality", "1", CVAR_ARCHIVE },
```
Comment (Aciz) [inline src/cgame/cg_main.cpp:1023]: Should this be on by default btw?

File: src/cgame/etj_strafe_quality_drawable.h [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+		double _totalFrames{ 0 };
+		double _goodFrames{ 0 };
+		double _strafeQuality{ 0 };
+
+		float _oldSpeed{ 0 };
+		int _team{ 0 };
+		mutable vec4_t _color;
+
+		// amount of digits to show on hud
+		static constexpr std::size_t _digits = 4;
```
Comment (Aciz) [inline src/cgame/etj_strafe_quality_drawable.h:40]: I assume the intention here is to show percentage with two decimals, however in cases where quality is < 10, you get three decimals instead, and on a theoretical 100% quality, only once decimal.

Comment (suburbski) [reply src/cgame/etj_strafe_quality_drawable.h:43]: No, the intention is to always have 4 digits in order to keep hud width consistent

**Comment by Aciz** [comment]
@isRyven good for merge, or is there something else still?

**Comment by isRyven** [comment]
lgtm

**Pull request merged**